### PR TITLE
Major feature: Stratification (e.g., taxonomy/function)

### PR DESCRIPTION
**Pull request by qiyunzhu**
This PR introduces a much demanded feature: **stratification**. This feature combines two classification systems in one analysis. A popular use case is taxonomic / functional profiling. Because functional genes typically form metabolic pathways within the same genome, analysis of metagenome should be informed by both taxonomy (i.e., genome) and function (i.e., gene) simultaneously. Previous, [HUMAnN2](http://huttenhower.sph.harvard.edu/humann) is well-known for performing this analysis.

The design of stratification in Woltka is as follows: First, run Woltka using one classification system (e.g., taxonomy), and choose to output read-to-taxon maps (using `--outmap <dir>`). Second, run Woltka again using another classification system (e.g., function), and take the previous maps to instruct stratification (using `--stratify <dir>`). This design is for flexibility and modularity.

The challenge of stratification is largely bonded by performance -- the program needs to match one map (sequence-to-taxonomy) and another (sequence-to-function), both are very huge files. The most efficient solution might be performing both analyses in memory in one run. However that will not guarantee flexibilty and modularity, therefore, for now, I choose to divide the analyses into two runs.

An example workflow is as follows (using bundled sample files):

Step 1: Taxonomic classification
```
woltka classify \
  -i align/bowtie2 \
  --map taxonomy/g2tid.txt \
  --nodes taxonomy/nodes.dmp \
  --names taxonomy/names.dmp \
  --rank phylum \
  --name-as-id \
  --outmap mapdir \
  -o taxonomy.biom
```

Step 2: Combined taxonomic/functional classification
```
woltka classify \
  -i align/bowtie2 \
  --coords function/coords.txt.xz \
  --map function/uniref.map.xz \
  --map function/go/process.tsv.xz \
  --map-as-rank \
  --rank process \
  --stratify mapdir \
  -o taxfunc.biom
```

The output feature table is like (note: the use of `|` follows HUMAnN2 convention):
```
#OTU ID S01     S02     S03     S04     S05
Actinobacteria|GO:0006071       2.0     0.0     0.0     0.0     0.0
Actinobacteria|GO:0006351       2.0     0.0     0.0     0.0     0.0
Actinobacteria|GO:0006412       0.0     0.0     0.0     0.0     2.0
Aquificae|GO:0000160    0.0     0.0     2.0     0.0     0.0
Aquificae|GO:0006011    0.0     0.0     2.0     0.0     0.0
Aquificae|GO:0006313    0.0     0.0     1.0     0.0     0.0
Bacteroidetes|GO:0006260        0.0     2.0     0.0     0.0     0.0
Bacteroidetes|GO:0006281        0.0     2.0     0.0     0.0     0.0
Bacteroidetes|GO:0006418        0.0     2.0     0.0     0.0     0.0
...
```

### CODE REVIEW

**Comment by qiyunzhu** [comment]
@gwarmstrong @fedarko @emikoifish @swandro

File: woltka/biom.py [hunk 1]
```diff
@@ -43,26 +43,28 @@ def profile_to_biom(profile, samples=None, tree=None, rankdic=None,
...
+        stratum, feature = key if isinstance(key, tuple) else (None, key)
+        obsrv, meta_ = feature, {}
         if namedic:
             name = namedic[feature] if feature in namedic else None
             if name_as_id:
-                names.append(name or feature)
+                obsrv = name or feature
             else:
                 meta_['Name'] = name or None
+        if stratum:
```
Comment (fedarko) [inline woltka/biom.py:65]: Are the `if namedic` and `if stratum` branches mutually exclusive? It looks like it, since `obsrv` is overwritten by the second branch.

Would suggest making this clearer -- e.g. by using `elif`, or at least adding a comment.

File: woltka/classify.py [hunk 2]
```diff
@@ -135,6 +135,36 @@ def count(matches):
     return res
 
 
+def count_strata(matches, strata):
+    """Stratify taxa in a map and count occurrences.
+
+    Parameters
+    ----------
+    matches : dict of str or dict
+        Query-to-taxon(a) map.
```
Comment (fedarko) [inline woltka/classify.py:144]: Not sure what `taxon(a)` means here

**Comment by qiyunzhu** [comment]
@fedarko Thank you very much for reviewing my code! I will address your comments asap!

File: doc/stratify.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Stratification
+
+It is a frequent need to combine two or more classification systems in one analysis. For example, in a typical shotgun metagenomics study, one not only needs the taxonomic composition and functional potential of whole samples, but also wants to know which functional genes are found in which taxonomic units. This is because in many scenarios, only proteins wrapped within the same cell can constitute metabolic pathways, and this information is only available when the _taxonomy_ and _function_ of each sequence are informed simultaneously.
+
+Woltka enables this analysis through a "**stratification**" function. A stratum (plural: strata) is a subset of data grouped under the same label. Woltka allows the user to label sequences using one classification system (e.g., taxonomy), then classify sequences under each label using another classification system (e.g., function). This design maximizes flexibility and modularity. It is not bond by particular classification levels or systems, but can also work with ecological niches, phenotypic features, and other classification methods that best address specific research goals.
```
[the remark this answers is not in the document]
Comment (qiyunzhu) [reply doc/stratify.md:5]: @fedarko Thanks for careful review of the documentation I wrote! I think the term "stratum" already has clear definition in statistics. See https://en.wikipedia.org/wiki/Stratified_sampling. I don't think there is need to re-define it. The expression "We define..." also makes the text less user-friendly (we are inventing jargons...). What do you think?

File: doc/stratify.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+Here is an example workflow, based on the commands and [sample data](../woltka/tests/data) introduced in the [quick-start guide](../README.md#example-usage), with merely two small modifications.
+
+Step 1: Taxonomic classification.
```
[the remark this answers is not in the document]
Comment (qiyunzhu) [reply doc/stratify.md:9]: Good point!

File: woltka/biom.py (hunk 1 above)
Comment (qiyunzhu) [reply woltka/biom.py:65]: They are NOT exclusive. `namedic` will replace taxon IDs with descriptive names or append names to the table. `stratum` will prefix taxon ID/name with stratum label. For examples:

- original: `2697049`
- with `namedic`: `SARS-CoV-2`
- with `stratum`: `spike|2697049`
- with `namedic` and `stratum`: `spike|SARS-CoV-2`

Comment (qiyunzhu) [reply woltka/biom.py:65]: I added some inline comments to make the code more clear.

File: woltka/classify.py (hunk 2 above)
Comment (qiyunzhu) [reply woltka/classify.py:144]: It means "taxon or taxa", analogous to "subject(s)".

**Comment by qiyunzhu** [comment]
@fedarko I just committed some of your suggestions and made some modifications upon your comments. Could you please take another look? Sorry for presenting code that is hard to follow, since the project is already large and it is favorable if we can meet and discuss the structure of the repository and the logic behind the workflow. I would like to do this and invite you to review more code, given you have interest and time. For now, George and I are doing this: Just merge code into the `master` branch, then review the `master` branch and submit issues, then tackle issues individually. This is because the codebase is large and a giant PR will be hard to follow. What do you think about this mechanism?

**Comment by qiyunzhu** [comment]
@fedarko @gwarmstrong I am going to click merge for now since I will need to inform users. Meanwhile I will submit a few issues so that reviewers can come back to address them later.

**Comment by fedarko** [comment]
@qiyunzhu sorry for the delay. I can take another look through the changes tomorrow.

I honestly don't have much time to help out regularly, but I'm happy to try to help with one-off things as needed :)

**Comment by qiyunzhu** [comment]
Hello @fedarko , thank you for your offer! The code is already merged but please feel free to leave comments in this PR. I will read and make responses /  edits.
